### ISSUE TITLE
softmax_loss -1 subtraction

### ISSUE THREAD

**Issue by mmuneeburahman** [mantasu/cs231n#10]
Here is a partial derivative jacobian matrix for softmax:
![image](https://github.com/mantasu/cs231n/assets/56772228/266874b5-b36f-49cb-b1cd-4bbca13a833a)
This simplifies to:
![image](https://github.com/mantasu/cs231n/assets/56772228/b203da3b-93b6-43f3-8d93-7967c223bcbc)
For diagonal values we have s . (1 - s) but you are doing s . (s -1) [here](https://github.com/mantasu/cs231n/blob/e8fb304bd9e0609ac758a07a1cc7101d21f4651b/assignment1/cs231n/classifiers/softmax.py#L43). Are both correct? Can someone explain?
For reference see [blog](https://towardsdatascience.com/derivative-of-the-softmax-function-and-the-categorical-cross-entropy-loss-ffceefc081d1).

**Comment by mantasu** [comment]
Just to clarify, for any sample $n$ when computing the loss, we only consider the predicted score of the class $c$ of interest, i.e., the class $c$ that is actually correct for that sample $n$:

$$L_n=\begin{cases}-\log\left(\frac{e^{\hat y_i}}{\sum_{i'} e^{\hat y_{i'}}}\right), & \text{when } i=c \qquad \\\\ 0, & \text{otherwise}\end{cases}$$

$$\nabla_{\hat y_i}L_n=\begin{cases}\left(\frac{e^{\hat y_i}}{\sum_{i'} e^{\hat y_{i'}}}\right)-1, & \text{when } i=c \\\\ \left(\frac{e^{\hat y_i}}{\sum_{i'} e^{\hat y_{i'}}}\right), & \text{otherwise}\end{cases}$$

The actual steps of arriving at the final derivative shown above are a bit more involved but you could easily solve that with pen and paper using the differentiation rules, like chain, logarithm, division etc., e.g., $\nabla_x \log(x)= \frac{1}{x}$ (we assume $\log=\text{ln}$).

Intuitively, that `-1` accounts for $\hat{y}_c$ appearing in the enumerator in the original calculation of $L_n$, which otherwise does not happen for any other $\hat{y}_i$ with $i \ne c$.

> **Extra note**: the full **cross-entropy loss** is calculated as the average negative log probability of the correct class across all your data points. For a single data point $n$, its negative log probability is just the "surprisal" of that event according to _information theory_.
